plugins/osdn/netid/controller/controller.go

```diff
@@ -0,0 +1,170 @@
+	tx := &tx{}
+	defer tx.Rollback()
+
+	var id uint
+	userRequestedVnid := false
+	value, err := netid.GetWantsVNID(ns)
+	if err == nil {
+		id = value
+		userRequestedVnid = true
+		netid.DeleteWantsVNID(ns)
+	} else if err != netid.ErrorVNIDNotFound {
+		return err
+	} else if _, err = netid.GetVNID(ns); err == nil {
+		// Nothing to do in this case
+		// VNID already allocated and namespace does not want new VNID
+		return nil
+	}
+
+	name := ns.ObjectMeta.Name
+	if userRequestedVnid {
+		// Only accept global VNID or VNIDs that are already allocated
+		if (id != netid.GlobalVNID) && !c.vnid.Has(id) {
+			return fmt.Errorf("requested netid %d not allocated", id)
+		}
+	} else {
+		// Do vnid allocation
+		if c.globalNamespaces.Has(name) {
+			id = netid.GlobalVNID
+		} else {
+			id, err = c.vnid.AllocateNext()
+			if err != nil {
+				return err
+			}
+			tx.Add(func() error { return c.vnid.Release(id) })
+		}
+	}
+	if err := netid.SetVNID(ns, id); err != nil {
+		return err
+	}
+
+	// TODO: could use a client.GuaranteedUpdate/Merge function
+	for i := 0; i < retryCount; i++ {
+		_, err := c.client.Update(ns)
+		if err == nil {
+			if userRequestedVnid {
+				glog.Infof("Updated netid %d for namespace %q", id, name)
+			} else {
+				glog.Infof("Assigned netid %d for namespace %q", id, name)
+			}
+			// commit and exit
+			tx.Commit()
+			return nil
+		}
+
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		if !errors.IsConflict(err) {
+			return err
+		}
+		newNs, err := c.client.Get(name)
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		newid, err := netid.GetVNID(newNs)
+		if (err == nil) && (newid != id) {
+			return fmt.Errorf("netid for namespace %q changed to %d on the flight, so can't assign netid %d", name, newid, id)
+		}
+
+		// try again
+		if err := netid.SetVNID(newNs, id); err != nil {
+			return err
+		}
+		ns = newNs
+	}
+
+	return fmt.Errorf("unable to allocate vnid for namespace %q after %d retries", name, retryCount)
+}
+
+// Delete releases the network id(VNID) for the namespace
+func (c *Allocation) Delete(ns *kapi.Namespace) error {
+	id, err := netid.GetVNID(ns)
+	if err != nil {
+		return nil
+	}
+
+	if id == netid.GlobalVNID {
+		return nil
+	}
+
+	// TODO: use namespace cache
+	// Do VNID Release if needed
+	nsList, err := c.client.List(kapi.ListOptions{})
+	if err != nil {
+		return err
+	}
+	name := ns.ObjectMeta.Name
+	for _, n := range nsList.Items {
+		// Skip current namespace
+		if n.ObjectMeta.Name == name {
+			continue
+		}
+		// Skip namespaces that doesn't match given id
+		if value, err := netid.GetVNID(&n); (err != nil) || (value != id) {
+			continue
+		}
+		// VNID is in use by other namespace
+		glog.Infof("Ignored releasing netid %d for namespace %q (still in use by %q)", id, name, n.ObjectMeta.Name)
+		return nil
+	}
+
+	err = c.vnid.Release(id)
+	if err != nil {
+		return err
+	}
+	glog.Infof("Released netid %d for namespace %q", id, name)
```

Comment:
Yeah, we continually run into deployments we need to debug where there's simply not enough logging to figure out what's going on.  We can't really ask the users to restart their cluster and reproduce the problem.  So we've been steadily adding logging to make it somewhat possible to debug issues.  We know we don't want logspam, so we're limiting the Info/v2/v3 stuff to things that happen rarely, like node add/remove, namespace add/remove, etc.